### ISSUE TITLE
problem with implementing DeepPhysX and Sofa

### ISSUE THREAD

**Issue by Abdelrahman-Alkhodary** [mimesis-inria/DeepPhysX.Sofa#3]
Hello @RobinEnjalbert 

I have built an environment trying to control soft arm using four cables with DeepPhysX library, but when I pull any of the cables the arm doesn't move. You can find my implementation [here](https://github.com/Abdelrahman-Alkhodary/DeepPhysX-Squid). Can you point me to where could be the problem.  

I have done this before using the softrobots plugins and it was working [SoftArm](https://github.com/Abdelrahman-Alkhodary/SoftArm-SOFA) and now i'm trying to use DeepPhysX library to generate some data. I started with a basic scene to make sure it was working before starting to generate some data.

**Comment by RobinEnjalbert** [comment]
Hi @Abdelrahman-Alkhodary ,

Thanks for using the project ! If I understand well, you would like to train a neural network to be able to control the soft arm given an effector target ?

Can you provide me more details about your problem ? Does your SOFA scene work outside of the scope of DeepPhysX ? What version of `SOFA` / `SofaPython3` / `SoftRobots` are you using ?

On my side, I use `SOFA` / `SofaPython3` `v22.12` versions (the latest release). As I never used `SoftRobots` before, so I freshly compiled the `v22.12` version of the plugin. I was able to launch your scene and the cable has a normal behavior (I guess so).
![Screenshot from 2023-02-23 11-08-36](https://user-images.githubusercontent.com/74538826/220877515-a7703c59-cbd4-4533-83fb-d1e0898a0989.png)

I made a [PR](https://github.com/Abdelrahman-Alkhodary/DeepPhysX-Squid/pull/1) on your Git to fix the deprecation warnings related to `v22.12` version of `SOFA` and to access the Data in a more convenient way, I hope this will help you !

**Comment by Abdelrahman-Alkhodary** [comment]
Hello @RobinEnjalbert 

Thanks for your reply and thanks for the pull request. Yes my goal is to train a neural network to act as an inverse model to control the soft arm. I'm using the latest binary Linux version of Sofa v22.12 and I have built caribou following the instructions in [here](https://caribou.readthedocs.io/en/latest/Building.html#/). The original scene I mentioned was built under an older version of Sofa I think it was v21.06 but when I import it to the current version of Sofa it works fine. 

I tried to change the code to use the DeepPhysX library and the problem I'm facing is that in the function onAnimateBeginEvent, when you press Animate one of the cables should be pulled and the arm should be moving but the arm doesn't respond and keeps falling. I tried your pull but the problem still exists on my side, did it work with you?

**Comment by RobinEnjalbert** [comment]
About the `Caribou` plugin, you don't necessarily need it to use `DeepPhysX.Sofa`, it is only required to run the demos that are implemented in the examples.

Okay so it behaves the same on my side, I thought that the deformations that you see above were actually generated with the cable. It is the really first time that I use `SoftRobots` so can you share a `SOFA` script that behaves normally out of the scope of `DeepPhysX` ?

**Comment by RobinEnjalbert** [comment]
This time I managed to have the right behavior (I guess).
![Screenshot from 2023-02-23 14-41-43](https://user-images.githubusercontent.com/74538826/220923916-e78fbce0-b46e-40af-a08a-2b49afbe3853.png)
Your scene was missing two components : a `FreeMotionAnimationsLoop` & a `GenericConstraintSolver`. I checked other examples to compare, this seems to solve your problem. 
The way to modify the cable constraint is a little bit different to the implement that I am used to, so your previous implementation of the `onAnimateBeginEvent` was probably correct.
I will to another PR on your Git with my working code.

**Comment by Abdelrahman-Alkhodary** [comment]
Thanks @RobinEnjalbert  for the help, it would have taken me more time to figure it the `AnimationLoop` problem as I thought it's added by default. 

I have another question about the difference between `step` function and `onAnimateBeginEvent` function, from what I understood that `step` function is several loops of `onAnimateBeginEvent` determined by the `simulations_per_step` variable in the `environementConfig`, am I right?. So, in your opinion which one could fit my project more to collect data, as for the arm when a cable is pulled it could take several animation loops for the end effector to reach its destination that could map to how much the cable is pulled. And the number of these animation loops would depend on how much the cable is pulled so it's a variable. Thanks for the help so far and sorry to bother.

**Comment by RobinEnjalbert** [comment]
Glad to see that it solved your problem !

You understood perfectly, the `step` function is calling `Sofa.Simulation.animate(root_node, dt)`, that triggers the animation events (`onAnimateBeginEvent` & `onAnimateEndEvent`). The `step` function is automatically called by `DeepPhysX` to produce data and the `simulations_per_step` variable can be set to run several steps of the simulation before really producing a training sample.

However, this number of steps in constant so it does not fit your problem since you will run a variable number of steps before producing a sample. The solution is to set the `simulations_per_step` to 1 and to call yourself `Sofa.Simulation.animate` while your simulation is not ready to produce data. You can implement your animation events like this:

- `onAnimateBeginEvent`: if you start a new sample, define a new cable pulling value, else do nothing;
- `onAnimateEndEvent`: check if the effector reached the destination; if it does, then define a training sample using `set_training_data` and start a new sample, else trigger a new step with `Sofa.Simulation.animate`.

This way, with a single call to `step`, you will be able to run a variable number of simulation steps while producing a single training sample.
